draco/analysis/hyforesbandpass.py

```diff
@@ -0,0 +1,520 @@
+        Returns
+        -------
+        vis: containers.HybridVisStream
+            DAYENU-filtered hybrid beamformed visibilities with
+            foreground residuals subtracted.
+        comp_bandpass: containers.VisBandpassCompensate
+                window-compensated gains along with the singular values
+                of the window, rank of the inverted window, and cutoff
+        """
+        # Distribute over products
+        hv.redistribute(["ra", "time"])   
+        source.redistribute(["ra", "time"])
+
+        # Validate that both hybrid beamformed visibilites match
+        if not np.array_equal(source.freq, hv.freq):
+            raise ValueError("Frequencies do not match for hybrid visibilities.")
+
+        if not np.array_equal(source.index_map["el"], hv.index_map["el"]):
+            raise ValueError("Elevations do not match for hybrid visibilities.")
+
+        if not np.array_equal(source.index_map["ew"], hv.index_map["ew"]):
+            raise ValueError("EW baselines do not match for hybrid visibilities.")
+
+        if not np.array_equal(source.index_map["pol"], hv.index_map["pol"]):
+            raise ValueError("Polarisations do not match for hybrid visibilities.")
+
+        if not np.array_equal(source.ra, hv.ra):
+            raise ValueError("Right Ascension do not match for hybrid visibilities.")
+
+        npol, nfreq, new, nel, ntime = hv.vis.local_shape ### Q:what is local_shape? Shape of the data distributed to a node, not the total shape?
+
+        # Step 4
+        # step 4: pseudo-invert the window:
+        # Singular values:  loop over pol: u[pol], s_val[pol], vh[pol] = LA.svd(W[pol,:,:]) # need the singular value to determine the pseduo-inverse cut off: cut_off = 3e-2 by default
+        # Pseudo-inverse:   loop over pol: W_pinv[pol,:,:], rank[pol] = LA.pinv(W[pol,:,:], atol = cut_off, return_rank = True) # return rank should be Nfreq - Nfreq_mask - Nfreq_delay_cut
+        # Unwindowed gains: loop over pol: g[pol, :] = W_pinv[pol,:,:].dot(y[pol, :])
+        # Compensate the window for the estimated bandpass gains
+        y = bp.bandpass[...]
+        W = bp.window[...]
+
+        s_val = np.zeros((npol, new, nfreq))
+        W_pinv = np.zeros_like(W)
+        rank = np.zeros((npol, new))
+        g = np.zeros_like(y)        
+
+        # if cutoff is 0.0, then do not compensate the window
+        if self.cutoff == 0.0:
+            g = y
+            self.log.debug("Skip compensating the window")
+        else:
+
+            self.log.debug("Start compensating the window")
+
+            for pp in range(npol):
+
+                for xx in range(new):
+
+                    u, s_val[pp, xx], vh = la.svd(W[pp, xx, :,:]) #Q6: is scipy.linalg imported as LA somewhere in the pipeline? #la lower case # create new module in draco analysis
+                    W_pinv[pp, xx, :, :], rank[pp, xx] = la.pinv(W[pp, xx, :, :], atol = self.cutoff, return_rank = True)
```

Comment:
If you don't need to save out `W_pinv`, this will be dramatically faster by using a `solve` instead, i.e. `g = la.solve(W, y)`. Off the top of my head, I don't know exactly what the structure of `W` is, but if is hermitian positive-definite, using `WL = la.cho_factor(w)` followed by `g = la.cho_solve(Wl, y)` is extremely fast. Otherwise, just using `la.solve` will still be significantly faster and more numerically stable than a pseudo-inverse.